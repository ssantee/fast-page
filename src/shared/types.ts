export type DomainEnv = {
  domain: string;
  account: string;
  region: string;
  adminDomain: string;
};
