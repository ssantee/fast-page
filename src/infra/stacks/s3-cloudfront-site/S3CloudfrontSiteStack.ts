import {
  aws_s3_deployment,
  CfnOutput,
  RemovalPolicy,
  Stack,
  StackProps,
  aws_route53 as route53,
  aws_ssm as ssm,
} from "aws-cdk-lib";
import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import { CloudFrontTarget } from "aws-cdk-lib/aws-route53-targets";
import { Construct } from "constructs";
import { CloudFrontToS3 } from "@aws-solutions-constructs/aws-cloudfront-s3";
import * as path from "node:path";
import { Distribution } from "aws-cdk-lib/aws-cloudfront";

interface WebPublicProps extends StackProps {
  deployEnv: string;
  assetsDir: string;
  certificateArnParamName: string;
  hzIdParamName: string;
  deployEnvDomain: string;
}

export class S3CloudfrontSiteStack extends Stack {
  public distribution: Distribution;

  constructor(scope: Construct, id: string, props: WebPublicProps) {
    super(scope, id);

    const certificateArn = ssm.StringParameter.valueForStringParameter(
      this,
      props.certificateArnParamName,
    );
    const hostedZoneId = ssm.StringParameter.valueForStringParameter(
      this,
      props.hzIdParamName,
    );

    const bucketRemovalPolicy =
      props.deployEnv === "dev" ? RemovalPolicy.DESTROY : RemovalPolicy.RETAIN;
    const autoRemoveObjects = props.deployEnv === "dev";

    const cfToS3 = new CloudFrontToS3(this, "fp-web-public", {
      bucketProps: {
        removalPolicy: bucketRemovalPolicy,
        autoDeleteObjects: autoRemoveObjects,
      },
      cloudFrontDistributionProps: {
        domainNames: [props.deployEnvDomain],
        certificate: Certificate.fromCertificateArn(
          this,
          "Cert",
          certificateArn,
        ),
      },
    });

    // create alias record in route53 for the CloudFront distribution
    new route53.AaaaRecord(this, "Alias", {
      zone: route53.PublicHostedZone.fromPublicHostedZoneAttributes(
        this,
        "HostedZoneRecord",
        {
          hostedZoneId: hostedZoneId,
          zoneName: props.deployEnvDomain,
        },
      ),
      target: route53.RecordTarget.fromAlias(
        new CloudFrontTarget(cfToS3.cloudFrontWebDistribution),
      ),
    });

    this.distribution = cfToS3.cloudFrontWebDistribution;

    new aws_s3_deployment.BucketDeployment(this, "DeployPlaceholderFile", {
      sources: [
        aws_s3_deployment.Source.asset(
          path.join(props.assetsDir, "/web-public"),
          {
            exclude: ["**", "!index.html"],
          },
        ),
      ],
      destinationBucket: cfToS3.s3BucketInterface,
      distribution: cfToS3.cloudFrontWebDistribution,
      // optional paths to invalidate in cache
      distributionPaths: ["/index.html"],
    });

    new CfnOutput(this, "BucketName", {
      value: cfToS3.s3Bucket?.bucketName || "NoBucketName",
    });

    new CfnOutput(this, "CFDistUrl", {
      value: `https://${
        cfToS3.cloudFrontWebDistribution.distributionDomainName || "NoCFDistUrl"
      }`,
    });
  }
}
