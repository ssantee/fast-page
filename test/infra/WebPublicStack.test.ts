import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { S3CloudfrontSiteStack } from "../../src/infra/stacks/s3-cloudfront-site/S3CloudfrontSiteStack";
import { describe } from "node:test";
import { DomainEnv } from "../../src/shared/types";

describe("S3CloudfrontSiteStack stack test suite", () => {
  let webPublicTemplate: Template;

  beforeAll(() => {
    const envs: { [env: string]: DomainEnv } = {
      dev: {
        account: "123456789012",
        region: "us-east-1",
        domain: "dev.pg.santee.cloud",
        adminDomain: "dev.pgadmin.santee.cloud",
      },
      prod: {
        account: "123456789012",
        region: "us-east-1",
        domain: "prod.pg.santee.cloud",
        adminDomain: "prod.pgadmin.santee.cloud",
      },
      root: {
        account: "123456789012",
        region: "us-east-1",
        domain: "pg.santee.cloud",
        adminDomain: "pgadmin.santee.cloud",
      },
    };

    const app = new cdk.App();
    const webPublicStack = new S3CloudfrontSiteStack(app, "WebPublicStack", {
      deployEnv: "prod",
      env: {
        account: "123456789012",
        region: "us-east-1",
      },
      description: "Web Public Stack",
      assetsDir: `${process.cwd()}/assets`,
      certificateArnParamName: "/account/fpCertificateArn",
      hzIdParamName: "/account/fpSubdomainHostedZoneId",
      deployEnvDomain: envs.prod.domain,
    });
    webPublicTemplate = Template.fromStack(webPublicStack);
  });

  test("construct produces 3 buckets", () => {
    // app bucket, access logging bucket, cloudfront logging and access logs buckets
    webPublicTemplate.resourceCountIs("AWS::S3::Bucket", 4);
  });

  test("s3 buckets have retention (prod), encryption, no public access", () => {
    webPublicTemplate.allResources("AWS::S3::Bucket", {
      DeletionPolicy: "Retain",
    });

    webPublicTemplate.allResourcesProperties("AWS::S3::Bucket", {
      PublicAccessBlockConfiguration: {
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      },
      BucketEncryption: {
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: "AES256",
            },
          },
        ],
      },
    });
  });

  test("Matches snapshot", () => {
    expect(webPublicTemplate).toMatchSnapshot();
  });
});
