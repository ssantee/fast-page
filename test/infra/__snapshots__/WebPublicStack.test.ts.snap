// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Outputs": {
    "BucketName": {
      "Value": {
        "Ref": "fpwebpublicS3Bucket09D00666",
      },
    },
    "CFDistUrl": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Fn::GetAtt": [
                "fpwebpublicCloudFrontDistribution89C368BD",
                "DomainName",
              ],
            },
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValuefpCertificateArnC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/fp/CertificateArn",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValuefpsubdomainHostedZoneIdC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/fp/subdomainHostedZoneId",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Alias325C5727": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "fpwebpublicCloudFrontDistribution89C368BD",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": {
          "Ref": "SsmParameterValuefpsubdomainHostedZoneIdC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "Name": "prod.fp.santee.cloud.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2d56e153cac88d3e0c2f842e8e6f6783b8725bf91f95e0673b4725448a56e96d.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "DeployPlaceholderFileAwsCliLayerB748471C",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "fpwebpublicS3Bucket09D00666",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "fpwebpublicS3Bucket09D00666",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeployPlaceholderFileAwsCliLayerB748471C": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3322b7049fb0ed2b7cbb644a2ada8d1116ff80c32dca89e6ada846b5de26f961.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "DeployPlaceholderFileCustomResource563C81A4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "fpwebpublicS3Bucket09D00666",
        },
        "DistributionId": {
          "Ref": "fpwebpublicCloudFrontDistribution89C368BD",
        },
        "DistributionPaths": [
          "/index.html",
        ],
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "3101de97d5cee5802833e7234d0b64b2e20fc4f7ffa938b22364994f3f872a80.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "fpwebpublicCloudFrontDistribution89C368BD": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W70",
              "reason": "Since the distribution uses the CloudFront domain name, CloudFront automatically sets the security policy to TLSv1 regardless of the value of MinimumProtocolVersion",
            },
          ],
        },
      },
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "prod.fp.santee.cloud",
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "FunctionAssociations": [
              {
                "EventType": "viewer-response",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "fpwebpublicSetHttpSecurityHeadersBF6376CB",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "TargetOriginId": "WebPublicStackfpwebpublicCloudFrontDistributionOrigin1B7E81B5E",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": {
              "Fn::GetAtt": [
                "fpwebpublicCloudfrontLoggingBucketFB4EA181",
                "RegionalDomainName",
              ],
            },
          },
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "fpwebpublicS3Bucket09D00666",
                  "RegionalDomainName",
                ],
              },
              "Id": "WebPublicStackfpwebpublicCloudFrontDistributionOrigin1B7E81B5E",
              "OriginAccessControlId": {
                "Fn::GetAtt": [
                  "fpwebpublicCloudFrontOac8939F9DE",
                  "Id",
                ],
              },
              "S3OriginConfig": {
                "OriginAccessIdentity": "",
              },
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "SsmParameterValuefpCertificateArnC96584B6F00A464EAD1953AFF4B05118Parameter",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "fpwebpublicCloudFrontOac8939F9DE": {
      "Properties": {
        "OriginAccessControlConfig": {
          "Description": "Origin access control provisioned by aws-cloudfront-s3",
          "Name": {
            "Fn::Join": [
              "",
              [
                "aws-cloudfront-s3-fp-wublic-",
                {
                  "Fn::Select": [
                    2,
                    {
                      "Fn::Split": [
                        "/",
                        {
                          "Ref": "AWS::StackId",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
          "OriginAccessControlOriginType": "s3",
          "SigningBehavior": "always",
          "SigningProtocol": "sigv4",
        },
      },
      "Type": "AWS::CloudFront::OriginAccessControl",
    },
    "fpwebpublicCloudfrontLoggingBucketAccessLog39BE823D": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W35",
              "reason": "This S3 bucket is used as the access logging bucket for another bucket",
            },
          ],
        },
      },
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "fpwebpublicCloudfrontLoggingBucketAccessLogPolicyEF3CD3B2": {
      "Properties": {
        "Bucket": {
          "Ref": "fpwebpublicCloudfrontLoggingBucketAccessLog39BE823D",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "fpwebpublicCloudfrontLoggingBucketAccessLog39BE823D",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "fpwebpublicCloudfrontLoggingBucketAccessLog39BE823D",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "ArnLike": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "fpwebpublicCloudfrontLoggingBucketFB4EA181",
                      "Arn",
                    ],
                  },
                },
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "logging.s3.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "fpwebpublicCloudfrontLoggingBucketAccessLog39BE823D",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "fpwebpublicCloudfrontLoggingBucketFB4EA181": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "fpwebpublicCloudfrontLoggingBucketAccessLog39BE823D",
          },
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "fpwebpublicCloudfrontLoggingBucketPolicy587BBFFC": {
      "Properties": {
        "Bucket": {
          "Ref": "fpwebpublicCloudfrontLoggingBucketFB4EA181",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "fpwebpublicCloudfrontLoggingBucketFB4EA181",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "fpwebpublicCloudfrontLoggingBucketFB4EA181",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "fpwebpublicS3Bucket09D00666": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "NoncurrentVersionTransitions": [
                {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "fpwebpublicS3LoggingBucket6AD22B06",
          },
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:a61b0210",
            "Value": "true",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "fpwebpublicS3BucketPolicy5C3CD2C6": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "F16",
              "reason": "Public website bucket policy requires a wildcard principal",
            },
          ],
        },
      },
      "Properties": {
        "Bucket": {
          "Ref": "fpwebpublicS3Bucket09D00666",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "fpwebpublicS3Bucket09D00666",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "fpwebpublicS3Bucket09D00666",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Condition": {
                "StringEquals": {
                  "AWS:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/",
                        {
                          "Ref": "fpwebpublicCloudFrontDistribution89C368BD",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudfront.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "fpwebpublicS3Bucket09D00666",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "fpwebpublicS3LoggingBucket6AD22B06": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W35",
              "reason": "This S3 bucket is used as the access logging bucket for another bucket",
            },
          ],
        },
      },
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "fpwebpublicS3LoggingBucketPolicyB364C819": {
      "Properties": {
        "Bucket": {
          "Ref": "fpwebpublicS3LoggingBucket6AD22B06",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "fpwebpublicS3LoggingBucket6AD22B06",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "fpwebpublicS3LoggingBucket6AD22B06",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "ArnLike": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "fpwebpublicS3Bucket09D00666",
                      "Arn",
                    ],
                  },
                },
                "StringEquals": {
                  "aws:SourceAccount": {
                    "Ref": "AWS::AccountId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "logging.s3.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "fpwebpublicS3LoggingBucket6AD22B06",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "fpwebpublicSetHttpSecurityHeadersBF6376CB": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "function handler(event) { var response = event.response; var headers = response.headers; headers['strict-transport-security'] = { value: 'max-age=63072000; includeSubdomains; preload'}; headers['content-security-policy'] = { value: "default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self'; object-src 'none'"}; headers['x-content-type-options'] = { value: 'nosniff'}; headers['x-frame-options'] = {value: 'DENY'}; headers['x-xss-protection'] = {value: '1; mode=block'}; return response; }",
        "FunctionConfig": {
          "Comment": "SetHttpSecurityHeadersc8a41598a73d6ab226777063920b869245daf78ec2",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": "SetHttpSecurityHeadersc8a41598a73d6ab226777063920b869245daf78ec2",
      },
      "Type": "AWS::CloudFront::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
